File: src/main/java/com/duduanan/leetcode/review/N0023_MergeKSortedLists.java
package com.duduanan.leetcode.review;

import java.util.PriorityQueue;

/**
 * 23. Merge k Sorted Lists
 * Hard
 * You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.
 * Merge all the linked-lists into one sorted linked-list and return it.
 */

class ListNode {
    int val;
    ListNode next;
    ListNode() {
    }
    ListNode(int val) {
        this.val = val;
    }
    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}
public class N0023_MergeKSortedLists {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);
        for (ListNode node : lists) {
            if (node == null) continue;
            pq.offer(node);
        }
        ListNode root = null;
        ListNode temp = null;
        while (!pq.isEmpty()) {
            ListNode tempNode = pq.poll();
            if (temp == null) {
                root = tempNode;
                temp = tempNode;
            } else {
                temp.next = tempNode;
                temp = temp.next;
            }
            if (tempNode.next != null) {
                pq.offer(tempNode.next);
            }
        }
        return root;
    }
}

File: c++/N0023_merge_k_sorted_Lists.cpp
#include  <queue>
#include <vector>
#include "list_node.h"
using namespace std;

class Solution {
    public:
        ListNode* mergeKLists(vector<ListNode*> &lists)
        {
             auto f = [](ListNode* l1, ListNode* l2) -> bool { return l1 -> val > l2 -> val;};
             priority_queue<ListNode*, vector<ListNode*>, decltype(f)> pq(f);  

             ListNode* dumpy = new ListNode;
             ListNode* curr = dumpy;
             
             for(auto l : lists)
                if(l) pq.push(l);
            while(pq.size() > 0)
            {
                ListNode* temp = pq.top();
                pq.pop();
                curr -> next = temp;
                curr  = curr -> next;
                temp = temp -> next;
                if(temp) pq.push(temp);
            }

            curr = dumpy -> next;
            delete dumpy;
            return curr;
        }
};
